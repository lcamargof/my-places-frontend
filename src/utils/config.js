export const MAPBOX_TOKEN = 'pk.eyJ1IjoibHVpc2t4IiwiYSI6ImNqZzc3OGhyOTFreG8zM2xvNDJ2MmE4amUifQ.-4-kEUuKIeCmvchZlLviKQ';
export const API_URL = process.env.NODE_ENV === 'development' ? 'http://localhost:5001' : '';
export const DEFAULT_COORDINATES = [-103.40932570000001, 20.6713195];