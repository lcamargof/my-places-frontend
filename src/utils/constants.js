export const ADD_ACTION = 'ADD';
export const EDIT_ACTION = 'EDIT';