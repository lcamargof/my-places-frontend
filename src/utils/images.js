const image = new Image();
image.src = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACQAAAAkCAQAAABLCVATAAABP0lEQVR4AeXUP0tCURjH8S8OhtKV/kwX13oPQatEEs3ZnhlFmXOt0VRzi70HQ0FoioiGcqjoDQi1pYHUcPX2NAU94D3nXB39fDd/j3fwgkwmjyJXtAkIaFNji2liS1Chg6D7oEyCGDJcIxE18XCU5h4xdEcKJ1XE0gUOlhFrPyxhVUdUfVo8ECCqGhbzDNQXGvgA+DT045nBaF2dP5PkT5IXta1hVFHHJbVtq+0Qo2N1nFNbTm1HGO2r46LaSmrbw2g18jea4lVtKxjNEQ59a1ma6vOQWSxuEVXAIy36iOoGq13EoR2sMvQQSz08HJwjls5w4vONGPrCx9EpYugEZx7vSERveMSwiURUIKYGMqQ6sWXpIug6ZCG+AoJugxFdqsdUGVn637/iEynGsMgngtBlgTHlCRmQh/GVOWCS/QJAbX2ObhdYKAAAAABJRU5ErkJggg==';
export const IMAGES = ['place', image];
