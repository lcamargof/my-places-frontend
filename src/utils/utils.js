export const getUserLocation = () => {
  return new Promise((resolve, reject) => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(position => (
        resolve([position.coords.longitude, position.coords.latitude])
      ), () => reject());
    } else {
      reject();
    }
  });
};