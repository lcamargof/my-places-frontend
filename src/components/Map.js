import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactMapboxGl, { Layer, Feature, Popup } from "react-mapbox-gl";
import { DEFAULT_COORDINATES, MAPBOX_TOKEN } from '../utils/config';
import Place from './Place';
import ConfirmDialog from './ConfirmDialog';
import PlaceFormDialog from './PlaceFormDialog';
import { ADD_ACTION, EDIT_ACTION } from '../utils/constants';

const MapBox = ReactMapboxGl({ accessToken: MAPBOX_TOKEN });

class Map extends Component {

  state = {
    location: DEFAULT_COORDINATES,
    zoom: [11],
    place: undefined,
    confirm: false,
    form: false,
    type: ADD_ACTION,
  };

  componentDidMount() {
    // Get current client position
    // It throw if it takes more than 10 seconds because if it happens after
    // It creates a weird transition
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(position => (
        this.setState({ location: [position.coords.longitude, position.coords.latitude] })
      ), err => console.log('err', err), { timeout: 10000 });
    }
  }

  componentWillReceiveProps(nextProps) {
    // React when the user filter a place using the search bar and set it active
    if (nextProps.place !== this.props.place && nextProps.place !== this.state.place) {
      this.setState({ place: nextProps.place, location: nextProps.place.location, zoom: [14] });
    }
  }

  handleClick = ({ feature }, place) => {
    this.setState({
      location: feature.geometry.coordinates,
      zoom: [14],
      place,
      type: EDIT_ACTION
    });
  };

  onToggleHover(cursor, { map }) {
    map.getCanvas().style.cursor = cursor;
  }

  onDrag = () => {
    this.setState({ place: undefined });
  };

  onDragEnd = ({ lngLat }, place) => {
    this.props.onUpdate(Object.assign({}, place, { location: [lngLat.lng, lngLat.lat] }));
  };

  addPlace = (evt, { lngLat }) => {
    this.setState({ place: { location: [lngLat.lng, lngLat.lat] }, type: ADD_ACTION });
    this.toggleForm(true);
  };

  handleEdit = () => {
    this.toggleForm(true);
  };

  handleRemove = () => {
    this.props.onDelete(this.state.place);
    this.setState({ place: undefined });
    this.toggleConfirm(false);
  };

  toggleForm = form => this.setState({ form });

  toggleConfirm = confirm => this.setState({ confirm });

  handleSubmit = (place) => {
    if (this.state.type === ADD_ACTION) {
      this.props.onAdd(place);
    } else {
      this.props.onUpdate(place);
    }

    this.setState({ place: undefined });
    this.toggleForm(false);
  };

  renderLayers = (places) => {
    const layers = places.map(p => (
      <Layer
        key={p.id}
        id={`${p.id}`}
        type="symbol"
        layout={{
          "icon-image": "rocket-15",
          "text-field": p.name,
          "text-offset": [0, 1.5],
          "icon-size": 1.8,
        }}>
        <Feature
          onMouseEnter={e => this.onToggleHover('pointer', e)}
          onMouseLeave={e => this.onToggleHover('', e)}
          coordinates={p.location}
          onClick={e => this.handleClick(e, p)}
          draggable
          onDragStart={this.onDrag}
          onDragEnd={e => this.onDragEnd(e, p)}
        />
      </Layer>
    ));

    return layers;
  };

  render() {
    const { location, place, zoom, type, confirm, form } = this.state;
    const { places } = this.props;

    return (
      <React.Fragment>
        <MapBox
          style="mapbox://styles/mapbox/streets-v9"
          center={location}
          zoom={zoom}
          onClick={this.onDrag}
          onDrag={this.onDrag}
          onContextMenu={this.addPlace}
          containerStyle={{
            height: "calc(100vh - 64px)",
            width: "100vw",
          }}>
          {
            this.renderLayers(places)
          }
          {
            place && place.name && (
              <Popup key={place.id} coordinates={place.location}>
                <Place
                  place={place}
                  onDelete={() => this.toggleConfirm(true)}
                  onEdit={this.handleEdit}
                />
              </Popup>
            )
          }
        </MapBox>
        <ConfirmDialog
          open={confirm}
          place={this.state.place}
          onConfirm={this.handleRemove}
          onClose={() => this.toggleConfirm(false)}
        />
        <PlaceFormDialog
          type={type}
          place={place}
          open={form}
          onSubmit={this.handleSubmit}
          onClose={() => this.toggleForm(false)}
        />
      </React.Fragment>
    );
  }
}

Map.propTypes = {
  places: PropTypes.array.isRequired,
  onAdd: PropTypes.func.isRequired,
  place: PropTypes.object,
  onDelete: PropTypes.func.isRequired,
  onUpdate: PropTypes.func.isRequired,
};

export default Map;