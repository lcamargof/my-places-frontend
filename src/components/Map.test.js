import React from 'react';
import { shallow } from 'enzyme';
import Map from './Map';
import { ADD_ACTION, EDIT_ACTION } from '../utils/constants';

jest.mock('mapbox-gl', () => ({
  Map: jest.fn(() => ({ on: jest.fn(), getCenter: jest.fn() }))
}));

describe('Component: Map', () => {
  let component = null;
  var places, onAdd, place, onUpdate, onDelete;

  beforeEach(() => {
    places = [
      { name: 'test', open: '00:00', close: '23:59', description: 'test', location: [10, 10] }
    ];
    place = undefined;
    onAdd = jest.fn();
    onDelete = jest.fn();
    onUpdate = jest.fn();
  });

  const buildComponent = () => {
    const props = {
      places, onAdd, place, onUpdate, onDelete
    };

    return shallow(<Map {...props} />);
  };

  describe('Render', () => {
    it('Should render correctly', () => {
      component = buildComponent();
    });
  });

  describe('Methods', () => {
    it('renderLayers: Should render the correct number of layers', () => {
      component = buildComponent();
      expect(component.instance().renderLayers(places).length).toBe(1);
    });

    it('handleSubmit: should call the correct method', () => {
      component = buildComponent();

      // ADD
      component.setState({ type: ADD_ACTION });
      component.instance().handleSubmit({});
      expect(onAdd).toHaveBeenCalledTimes(1);
      // EDIT
      component.setState({ type: EDIT_ACTION });
      component.instance().handleSubmit({});
      expect(onUpdate).toHaveBeenCalledTimes(1);
    });

    it('Should set place on prop update', () => {
      component = buildComponent();
      const place = {};
      component.setProps({ place });
      expect(component.state().place).toBe(place);
    });
  });
});