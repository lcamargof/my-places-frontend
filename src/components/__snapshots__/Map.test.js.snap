// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Map Snapshots Should render correctly 1`] = `
<React.Fragment>
  <ReactMapboxGl
    bearing={0}
    center={
      Array [
        -103.40932570000001,
        20.6713195,
      ]
    }
    containerStyle={
      Object {
        "height": "calc(100vh - 64px)",
        "width": "100vw",
      }
    }
    movingMethod="flyTo"
    onClick={[Function]}
    onContextMenu={[Function]}
    onDrag={[Function]}
    onStyleLoad={[Function]}
    pitch={0}
    style="mapbox://styles/mapbox/streets-v9"
    zoom={
      Array [
        11,
      ]
    }
  >
    <EnhancedLayer
      id="undefined"
      layout={
        Object {
          "icon-image": "rocket-15",
          "icon-size": 1.8,
          "text-field": "test",
          "text-offset": Array [
            0,
            1.5,
          ],
        }
      }
      type="symbol"
    >
      <Feature
        coordinates={
          Array [
            10,
            10,
          ]
        }
        draggable={true}
        onClick={[Function]}
        onDragEnd={[Function]}
        onDragStart={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      />
    </EnhancedLayer>
  </ReactMapboxGl>
  <ConfirmDialog
    onClose={[Function]}
    onConfirm={[Function]}
    open={false}
    place={Object {}}
  />
  <WithStyles(PlaceFormDialog)
    onClose={[Function]}
    onSubmit={[Function]}
    open={false}
    type="ADD"
  />
</React.Fragment>
`;
